### PR TITLE
CORTX-28270 Updated readme and other architectural docs.

### PR DESCRIPTION
**Pull request by seagate-sarang-sawant**
# Problem Statement
-Update Readme and other docs for better usability.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
-Typically a member contributing to test framework would follow the review process as follows:
-	1. We are following the concept of upstream and downstream where commits happen on your forked repository (downstream)
-	2. Then you can raise a PR to merge it to Seagate Cortx-Test repository (Upstream)
-	3. Moderators of Cortx-Test can create server side feature branch if multiple developers are working on same feature branch
-	4. Team member should be able to check-in and raise the PR to upstream even if they have read-only access to Seagate Repositories using this process
+Typically, a member contributing to test framework would follow the review process as follows:
+1. Commits happen on your forked repository(origin).  
+2. Then you can raise a PR to merge it to Seagate Cortx-Test repository (Upstream)
+3. Moderators of Cortx-Test can create server side feature branch if multiple developers are working on same feature branch
+4. Team member should be able to check-in and raise the PR to upstream even if they have read-only access to Seagate Repositories using this process
```
Comment (apoorva-rao) [inline README.md:22]: Is point no. 4 an extension of the point no. 2? If yes, maybe we can re-word the 2nd point as follows -

2. Then you can raise a PR to merge it to Seagate Cortx-Test repository (Upstream). PRs can be raised even if you have read-only access to Seagate Repositories.

File: README.md [hunk 2]
```diff
@@ -160,134 +163,294 @@ make clean --makefile=scripts/s3_tools/Makefile
 Cortx-test uses MongoDB as backend to store Cortx setup details. These details, stored in MongoDB, are specific
 to the setup itself. The purpose of this setup is to do automatic config generation
 based on the setup. Not all values are mandatory and only applicable values needs to be filled in vm environment. A sample template is as shown below. This template is feed to database and pulled when developer will run test automation with test runner. The pulled templates merges with static yaml files to build the CMN_CFG and other component level configs.
```
Comment (apoorva-rao) [inline README.md:165]: Minor typo corrections for "fed" and "merge"

```suggestion
based on the setup. Not all values are mandatory and only applicable values needs to be filled in vm environment. A sample template is as shown below. This template is fed to the database and pulled when developer will run test automation with test runner. The pulled templates merge with static yaml files to build the CMN_CFG and other component level configs.
```

File: docs/Architectural-Overview.md [hunk 3]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+* Autobot: A program which sets up necessary environment and pre-requisites to run test automation pointing to a single target. 
+
+* Test Execution Framework: Loosely coupled test framework module responsible for collecting test metadata from Jira and executing optimized execution ticket provided by 
+
+Kafka. 
+
+* Test Executor framework: Same as a test execution framework. 
+
+* Dist Runner: Test execution framework module loosely couple for test framework and consumer module of execution framework.  It is responsible for collecting test metadata from Jira and creating an optimized test execution plan as per Jira Test Plans.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:147]: Minor typo correction - 
```suggestion
* Dist Runner: Test execution framework module loosely coupled for test framework and consumer module of execution framework.  It is responsible for collecting test metadata from Jira and creating an optimized test execution plan as per Jira Test Plans. 
```

File: docs/Architectural-Overview.md [hunk 4]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+
+# Framework Components: 
+
+ 
+
+The following diagram shows all components/modules present in framework.  
+
+
+*   Master autobot: This module reads test execution data from test management tool. From this test execution data, it identifies which tests need to be executed. It performs analytics on test data and divides test cases into different chucks. Those chucks are executed by individual execution autobot.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:174]: Typo correction for 'chucks' - 
```suggestion
*   Master autobot: This module reads test execution data from test management tool. From this test execution data, it identifies which tests need to be executed. It performs analytics on test data and divides test cases into different chunks. Those chunks are executed by individual execution autobot. 
```

File: docs/Architectural-Overview.md [hunk 5]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+*   Logging: Logging support is provided by framework. It captures all the required logs. After test executions, framework stores all logs to nfs share, so that it can be accessible to user at any time. 
+
+ 
+
+*   Health Check: This module checks the health of target system. If the target system is in bad condition, then there is no use of using that system for test execution. So, before starting any test, framework checks the health of the target system using this module. If the system is not healthy, then framework schedules the test execution on another healthy system. 
+
+ 
+
+*   Config Management: All test data which required for test execution is managed by this module. Target specific config data is stored in MongoDB, test suite specific config data is stored in Jira/database, and test specific data is stored locally. This module captures and aggregates config data and makes that data/config maps available to all tests during execution.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:202]: Minor corrections - 
```suggestion
*   Config Management: All the test data which is required for test execution is managed by this module. Target specific config data is stored in MongoDB, test suite specific config data is stored in Jira/database, and test specific data is stored locally. This module captures and aggregates config data and makes that data/config maps available to all tests during execution. 	 
```

File: docs/Architectural-Overview.md [hunk 6]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+ 
+
+ 
+
+ 
+
+## Framework Deployment Flow: 
+
+This framework can be deployed and used in two ways. In multi auto bot setup mode and single auto bot setup mode.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:224]: Maybe we can remove unintentional whitespace -
```suggestion
This framework can be deployed and used in two ways. In multi autobot setup mode and single autobot setup mode. 
```

File: docs/Architectural-Overview.md [hunk 7]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+ 
+
+## Framework Deployment Flow: 
+
+This framework can be deployed and used in two ways. In multi auto bot setup mode and single auto bot setup mode. 
+
+ 
+
+## Multi Auto bot's setup:
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:228]: Same as 224 - 
```suggestion
## Multi Autobot's setup: 
```

File: docs/Architectural-Overview.md [hunk 8]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+ 
+
+ 
+
+sequence 1 to 6 is explained below in points. 
+
+Jenkins/Scheduler triggers a job which supplies parameters like targets and config to distributed executor i.e., master autobot. 
+
+Master Autobot runs in distributed mode and creates an execution plan which determines the sequence of test execution and which tests can be run in parallel. This execution plan entries are feed to Kafka topic cortx-test-jobs in JSON format.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:238]: Minor typos - 
```suggestion
Master Autobot runs in distributed mode and creates an execution plan which determines the sequence of test execution and which tests can be run in parallel. This execution plan's entries are fed to Kafka topic cortx-test-jobs in JSON format. 
```

File: docs/Architectural-Overview.md [hunk 9]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+ 
+
+sequence 1 to 6 is explained below in points. 
+
+Jenkins/Scheduler triggers a job which supplies parameters like targets and config to distributed executor i.e., master autobot. 
+
+Master Autobot runs in distributed mode and creates an execution plan which determines the sequence of test execution and which tests can be run in parallel. This execution plan entries are feed to Kafka topic cortx-test-jobs in JSON format. 
+
+Test runners (single autobot) are running in distributed mode on one or more virtual machines which have Kafka consumer. They belong to the same consumer group, and it guaranteed that a test will be executed only once by any of the test runners.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:240]: Minor suggestion - 
```suggestion
Test runners (single autobot) are running in distributed mode on one or more virtual machines which have Kafka consumer. They belong to the same consumer group, and it is guaranteed that a test will be executed only once by any of the test runners. 
```

File: docs/Architectural-Overview.md [hunk 10]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+## Benefits: 
+
+Can run on commodity H/W and light weight VMs. The run can be started on a single client VM for multiple targets. (Storage efficient) 
+
+Framework will start parallel execution on multiple targets; depending on health of target, it will reschedule test execution for errored or failed tests. This will ensure complete test execution of identified test plan.  
+
+Framework will have parallel execution over a single target also which will reduce the time required for test execution. (Time efficient) 
+
+Maximum failures will get during the first few hours of execution.
```
Comment (apoorva-rao) [inline docs/Architectural-Overview.md:286]: Maybe we can reword it to "Parallel execution will reduce the time required to uncover failures during testing"

**Review by apoorva-rao — CHANGES_REQUESTED** [review]
@seagate-sarang-sawant Please see my inline comments for some suggestions.

File: README.md (hunk 1 above)
Comment (pranav-sahasrabudhe) [reply README.md:22]: I am OK with this rewording or point#4 can be added as a footnote as well instead of separate bullet. Not a major change, good to have.

File: docs/Architectural-Overview.md [hunk 11]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
 ```
 
 # Hardware Configuration for test automation setup
-VMs or Physical machines with 16-32 GB RAM and 8-16 logical cores would be a decent configuration for a test machine. The deployment view [Test Execution Deployment](Test-Execution-Deployment-View.md) discusses more about the deployment structure. The least requirement is that there should be a network connectivity of these Cloud Form Vms or physical machines to the Cortx setups in labs.
+VMs or Physical machines with Linux flavor having 16-32 GB RAM and 8-16 logical cores would be a decent configuration for a test machine. Minimal test machine configuration is 4 GB RAM and 2 logical cores  which is suitable for development environment. The deployment view [Test Execution Deployment](Test-Execution-Deployment-View.md) discusses more about the deployment structure. The least requirement is that there should be a network connectivity of these Cloud Form Vms or physical machines to the Cortx setups in labs or public cloud services like AWS EC2 .
```
Comment (pranav-sahasrabudhe) [inline docs/Architectural-Overview.md:133]: IMO, We can remove the references to CloudForm VMs etc. specific to Seagate.

File: docs/Architectural-Overview.md [hunk 12]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+* Test Executor framework: Same as a test execution framework. 
+
+* Dist Runner: Test execution framework module loosely couple for test framework and consumer module of execution framework.  It is responsible for collecting test metadata from Jira and creating an optimized test execution plan as per Jira Test Plans. 
+
+* Chaos and destructive testing framework: A sub framework responsible for implementing libraries to generate destructive test scenarios and chaos in Target deployments. 
+
+* Object Store Data Integrity framework: A sub framework which stores s3 client state and provides library to test versioned object Store data integrity interleaved across object store version upgrade and after object store downtime.  
+
+* HA Tests: High availability test cases which tests Consistency, Availability and Reliability of Object Store.
```
Comment (pranav-sahasrabudhe) [inline docs/Architectural-Overview.md:153]:
```suggestion
* HA Tests: High availability test cases which test Consistency, Availability and Reliability of the Object Store. 
```

File: docs/Architectural-Overview.md [hunk 13]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+* Object Store Data Integrity framework: A sub framework which stores s3 client state and provides library to test versioned object Store data integrity interleaved across object store version upgrade and after object store downtime.  
+
+* HA Tests: High availability test cases which tests Consistency, Availability and Reliability of Object Store. 
+
+* Test Client: A separate VM or Container configured with Python Environment and prerequisites Linux and Python site packages to run automated test cases.      
+
+
+# Introduction: 
+
+Cortx-Test is distributed test execution framework comprising multiple test runners called autobots. These autobots runs heterogeneous tests of several types (functional, limits, scale, HA tests) on different targets in optimized manner and collect test results at central database (MongoDB). The test types supported are component or system level, functional tests, load tests, security tests, etc.
```
Comment (pranav-sahasrabudhe) [inline docs/Architectural-Overview.md:160]:
```suggestion
Cortx-Test is distributed test execution framework comprising of multiple test runners called autobots. These autobots run heterogeneous tests of several types (functional, limit, scale and HA tests) on different targets in optimized manner and collect test results at central database (MongoDB). The test types supported are component tests or system level, functional, load, security tests, etc.  
```

File: docs/Architectural-Overview.md [hunk 14]
```diff
@@ -125,7 +130,166 @@ Its a REST service which provides APIs to stores build wise test results and agg
...
+
+# Background: 
+
+For QA certification of an enterprise product, it is always a challenge to execute the test universe within a day especially since the test universe has hundreds of thousands of tests cases of several types including component, system, and performance tests. Job schedulers and CI tools like Jenkins solve this problem to an extent by providing functionality to executing multiple tests on different machines, however it is always hard to monitor numerous test executions and deploy and manage targets and complete certification of a build within a day. The algorithms and methods mentioned in the paper explain design and implementation of test executor framework which solves this problem with automatic test configuration management, parallel test execution, reporting and storage efficiency.
```
Comment (pranav-sahasrabudhe) [inline docs/Architectural-Overview.md:164]:
```suggestion
For QA certification of an enterprise product, it is always a challenge to execute the test universe within finite timeframe (typically a day) especially since the test universe has thousands of test cases of several types including component, system and performance tests. Job schedulers and CI tools like Jenkins solve this problem to an extent by providing functionality to execute multiple tests on different machines, however it is hard to monitor numerous test executions and deploy and manage targets and complete certification of a build within a day. The algorithms and methods mentioned in the paper explain design and implementation of test executor framework which solves this problem with automatic test configuration management, parallel test execution, reporting and storage efficiency.       
```

**Review by pranav-sahasrabudhe — CHANGES_REQUESTED** [review]
Added a few comments.

File: README.md (hunk 1 above)
Comment (seagate-sarang-sawant) [reply README.md:22]: anyway the repo is public now.

File: README.md (hunk 2 above)
Comment (seagate-sarang-sawant) [reply README.md:165]: correct verbiage and typos.

File: docs/Architectural-Overview.md (hunk 11 above)
Comment (seagate-sarang-sawant) [reply docs/Architectural-Overview.md:133]: I have added a generic VM.

File: docs/Architectural-Overview.md (hunk 4 above)
Comment (seagate-sarang-sawant) [reply docs/Architectural-Overview.md:174]: fixed

**Review by apoorva-rao — APPROVED** [review]

**Review by pranav-sahasrabudhe — APPROVED** [review]
Looks good to me.
